File: go/two_pointers/11_container_with_most_water.go
package main

import (
	"fmt"
)

func main() {
	// Test cases

	// result: 49
	fmt.Println(maxArea([]int{1, 8, 6, 2, 5, 4, 8, 3, 7}))

	// result: 1
	fmt.Println(maxArea([]int{1, 1}))
}

func maxArea(height []int) int {
	i := 0
	k := len(height) - 1

	max := 0
	for i < k {
		b := k - i
		h := height[i]
		if h > height[k] {
			h = height[k]
		}
		a := b * h

		if a > max {
			max = a
		}

		if height[i] == height[k] {
			k--
			continue
		}
		if height[i] == h {
			i++
		} else {
			k--
		}
	}
	return max
}

File: ruby/two_pointers/11_container_with_most_water.rb
# @param {Integer[]} height
# @return {Integer}
def max_area(height)
  i = 0
  k = height.size - 1

  max = 0

  while i < k
    b = k - i
    h = [height[i], height[k]].min
    a = b * h

    max = a > max ? a : max

    k -= 1 and next if height[k] <= h
    i += 1 and next if height[i] == h
  end

  max
end

def test
  # Test cases

  # result: 49
  puts max_area [1, 8, 6, 2, 5, 4, 8, 3, 7]

  # resut: 1
  puts max_area [1, 1]
end

test
